assembly/index.ts

```diff
@@ -0,0 +1,203 @@
+
+import { base64, runtime_api } from "near-runtime-ts";
+import { JSONEncoder as _JSONEncoder, JSON } from "assemblyscript-json";
+import { u128 } from "bignum";
+
+// Runtime functions
+// tslint:disable: no-unsafe-any
+
+@global
+class JSONEncoder extends _JSONEncoder {
+  encode<T>(name: string, val: T): void {
+    encode<T, JSONEncoder>(val, name, this);
+  }
+}
+
+type Usize = u64;
+//@ts-ignore
+@global
+function read_register(register_id: Usize, ptr: Usize): void {
+  runtime_api.read_register(register_id, ptr);
+}
+//@ts-ignore
+@global
+function register_len(register_id: Usize): Usize {
+  return runtime_api.register_len(register_id);
+}
+
+//@ts-ignore
+@global
+function input(register_id: Usize): void {
+  runtime_api.input(register_id);
+}
+//@ts-ignore
+@global
+function value_return(value_len: Usize, value_ptr: Usize): void {
+  runtime_api.value_return(value_len, value_ptr);
+}
+//@ts-ignore
+@global
+function panic(): void {
+  runtime_api.panic();
+}
+
+//@ts-ignore
+@global
+function getInput(): JSON.Obj { 
+  // Reading input bytes.
+  input(0);
+  let json_len = register_len(0);
+  if (json_len == U32.MAX_VALUE) {
+    panic();
+  }
+  let json = new Uint8Array(json_len as u32);
+  //@ts-ignore
+  read_register(0, <usize>json.buffer);
+  return <JSON.Obj> JSON.parse(json);
+}
+
+
+//@ts-ignore
+@global
+function encode<T, Output = Uint8Array>(value: T, name: string | null = "", encoder: JSONEncoder = new JSONEncoder()): Output {
+  if (isBoolean<T>()) {
+    //@ts-ignore
+    encoder.setBoolean(name, value);
+  } else if (isInteger<T>()) {
+    if (value instanceof i64 || value instanceof u64) {
+      //@ts-ignore
+      encoder.setString(name, value.toString());
+    } else {
+    //@ts-ignore
+      encoder.setInteger(name, value);
+    }
+    //@ts-ignore
+  } else if (value == <T>null) {
+    encoder.setNull(name);
+  } else if (isString<T>()) {
+    //@ts-ignore
+    encoder.setString(name, value);
+  } else if (isReference<T>()) {
+    if (isArrayLike<T>(value)) {
+      if (value instanceof Uint8Array) {
+        //@ts-ignore
+        encoder.setString(name, base64.encode(<Uint8Array> value));
+      } else {
+        encoder.pushArray(name);
+        for (let i: i32 = 0; i < value.length; i++) {
+          //@ts-ignore
+          encode<valueof<T>, JSONEncoder>(value[i], null, encoder);
+        }
+        encoder.popArray();
+      }
+    } else { // Is an object
+      if (value instanceof u128) {
+        encoder.setString(name, value.toString());
+      } else {
+        //@ts-ignore
+        value._encode(name, encoder);
+      }
+    }
+  } else {
+    throw new Error("Encoding failed");
```

Comment:
should this give more details on why encoding failed? Also how does this work, AssemblyScript now has proper exceptions with stack traces?

Reply:
Given the nesting of conditionals this should be a dead branch as all the possible types are covered.  And the error is just a wrapper around an abort, which provides a message and then executes the `unreachable` instruction that throws an exception to the host to deal with.